designdocs/recovery_system.md

```diff
@@ -0,0 +1,505 @@
+
+#### 3. Off-Chain Validation (Email Ownership)
+
+Before proceeding, the recovery provider must ensure the person requesting recovery controls the email address on file. This mitigates attacks where someone finds a lost recovery phrase but doesn’t own the email, for example. The provider performs an email challenge:
+
+- The provider generates a one-time random verification code (e.g. a 6-8 digit numeric code or a secure random token) and sends it to the user’s email address on record. This is typically done by sending an email with the code or a verification link containing the token.
+- The user must retrieve this code from their email and provide it back to the recovery provider (e.g. enter it on the website). This step proves they have access to that email inbox. The codes should be short-lived (e.g. 5 or 10 minutes expiry) and single-use. Phishing protection: The email should clearly state the purpose (“Your Frequency account recovery code is…”) and not include any clickable link that could be hijacked; a plain code that the user manually copies is often considered safer.
+- The provider verifies the code matches and is within the allowed time. If correct, email ownership is confirmed. If not, recovery is aborted (the provider will not proceed to add a new key).
+
+By performing this two-factor validation – “something you have (recovery code) and something you control (email account)” – the system greatly reduces the chance of unauthorized takeover. An attacker would need both the code phrase and access to the user’s email to succeed.
+
+#### 4. New Key Generation
+
+Once validation passes, the provider generates a new Control Key for the MSA. There are two approaches:
+
+- User-specified key: If the user is somewhat technical, they might generate a new key pair on their own device (e.g. via a wallet app) and only give the public key to the provider. This is ideal for security, as the private key never leaves the user. The recovery provider would ask “Please paste a new public key to add to your account.” However, this requires the user to know how to create a key pair, which may not be user-friendly in an emergency.
+- Provider-generated key: The recovery provider can create a new key pair on behalf of the user. For instance, it generates a new Ed25519, Sr25519, or Ethereum key (compatible with Frequency). The provider will then securely transmit the private key to the user after the on-chain update. This approach is simpler for the user but requires trust that the provider deletes its copy. In practice the provider could generate a mnemonic and show it to the user over a secure TLS connection for them to copy, minimizing its own storage of the key.
+
+This design recommends giving the user the option, but in either case, a public key for the new control key is decided at this step.
```

Comment:
I would say the choice of recovery provider dictates this interaction. We want to make it possible for a custodial wallet to be a recovery provider, in which case the user would not receive the private key at all.